lib/srv/app/server.go

```diff
@@ -802,8 +803,36 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		// Covert trace error type to HTTP and write response, make sure we close the
 		// connection afterwards so that the monitor is recreated if needed.
 		code := trace.ErrorToCode(err)
+
+		// Return a custom message for device trust errors.
+		var text string
+		if errors.Is(err, services.ErrTrustedDeviceRequired) {
+			// Try to guess the app from the host for a better error messages.
+			hostOnly, _, _ := net.SplitHostPort(r.Host)
+			appName := "yourapp"
+			for _, app := range s.getApps() {
+				if app.GetPublicAddr() == hostOnly || app.GetPublicAddr() == r.Host {
+					appName = app.GetName()
+					break
+				}
+			}
+
+			text = fmt.Sprintf(`Access to this app requires a trusted device.
+
+Try running:
+
+tsh proxy app %s -p 8888
+
+and then accessing the app via http://localhost:8888
+
+See https://goteleport.com/docs/access-controls/device-trust/device-management/#troubleshooting for help.
+`, appName)
+		} else {
+			text = http.StatusText(code)
+		}
+
```

Comment:
suggestion: pull this out into a function. I don't really want to read device trust error handling code in the main ServeHTTP for app access. You can avoid mutating the `text` variable by just returning it too